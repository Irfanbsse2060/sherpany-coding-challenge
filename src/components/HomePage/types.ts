export interface UserSearchInterface {
    name: string
}
