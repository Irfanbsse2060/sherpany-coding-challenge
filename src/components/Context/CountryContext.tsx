import React from 'react'

type Action = {type: 'update', payload: string[]}
type Dispatch = (action: Action) => void
type State = {countries: string[]}
type CountryProviderProps = {children: React.ReactNode}


const CountryStateContext = React.createContext<State | undefined>(undefined)
const CountryDispatchContext = React.createContext<Dispatch | undefined>(
    undefined,
)

function countryReducer(state: State, action: Action) {
    switch (action.type) {
        case 'update': {
            return {countries: action.payload}
        }
        default: {
            throw new Error(`Unhandled action type: ${action.type}`)
        }
    }
}



function CountryProvider({children}: CountryProviderProps) {
    const [state, dispatch] = React.useReducer(countryReducer, {countries: []})
    return (
        <CountryStateContext.Provider value={state}>
        <CountryDispatchContext.Provider value={dispatch}>
            {children}
            </CountryDispatchContext.Provider>
            </CountryStateContext.Provider>
    )
}


function useCountryState() {
    const context = React.useContext(CountryStateContext)
    if (context === undefined) {
        throw new Error('useCountState must be used within a CountProvider')
    }
    return context
}
function useCountryDispatch() {
    const context = React.useContext(CountryDispatchContext)
    if (context === undefined) {
        throw new Error('useCountDispatch must be used within a CountProvider')
    }
    return context
}


export {CountryProvider, useCountryState, useCountryDispatch}
