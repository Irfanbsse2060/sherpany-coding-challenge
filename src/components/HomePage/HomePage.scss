@import "../../styles/variables";

.main-page {
  display: grid;
  grid-template-columns: 1fr;
  grid-row-gap: 2rem;
  grid-column-gap: 2rem;
  padding: 2rem;

  &__users-list{
    overflow-x:auto;
    i {
      font-weight: lighter;
    }
  }
  .thumbnail {
    border-radius: 24px;
    width: 48px;
    height: 48px;
  }
}
