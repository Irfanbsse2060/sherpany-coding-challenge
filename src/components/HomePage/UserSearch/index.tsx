import React, {ChangeEvent, FormEvent, MouseEvent} from "react";

import './UserSearch.scss'
import {Button, ButtonType, TextField} from '../../UI'
import {UserSearchInterface} from "../types";

interface formProps {
    onSubmit: (userSearch:UserSearchInterface) => void
    onClearFilter: () => void
}

export default function UserSearch({onSubmit, onClearFilter}: formProps) {
    const initialState = {name: ''}
    const [state, setState] = React.useState(initialState)
    const formRef = React.useRef<HTMLFormElement>(null);

    const handleChange = (event: ChangeEvent<HTMLInputElement | HTMLTextAreaElement>) => setState({
        ...state,
        [event.target.name]: event.target.value
    })

    const {name} = state

    const handleSubmit = (event: FormEvent<HTMLFormElement> | MouseEvent<HTMLButtonElement>) => {
        event.preventDefault()
        if (formRef && formRef.current) {
            if (formRef.current.checkValidity()) {
                onSubmit(state)
            } else
                formRef.current.reportValidity();
        }
    }

    return (
        <form ref={formRef} onSubmit={handleSubmit} className='user-search'>
            <TextField label='Name' name='name' value={name} required={true} handleChange={handleChange}/>
            <div className='user-search__action'>
                <Button type={ButtonType.primary} handleClick={handleSubmit} text={'Search'}/>
                <Button type={ButtonType.secondary} handleClick={()=>{
                    setState(initialState)
                    onClearFilter()
                }} text={'Clear search'}/>
            </div>
        </form>
    )


}
