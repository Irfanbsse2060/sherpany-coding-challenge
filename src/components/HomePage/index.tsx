import React, {useEffect, useState} from 'react';
import InfiniteScroll from 'react-infinite-scroll-component';
import './HomePage.scss'
import UserSearch from './UserSearch'
import {
    Table
} from '../UI';
import {useCountryState} from '../Context/CountryContext'

import {filterUserByFirstNameAndLastName} from '../../utils/userUtil'
import {UserSearchInterface} from "./types";

interface User {
    email: string,
    gender: string,
    nat: string,
    name: {
        first: string,
        last: string
    },
    location: {
        country: string
    },
    picture: {
        thumbnail:string
    }
}

function HomePage() {
    const [userList, updateUserList] = useState<User[]>([])
    const [hasMoreUsers, updateHasMoreUsers] = useState(true)
    const [page, updatePage] = useState(1)
    const [filteredUserList, updateFilteredUserList] = useState<User[]>([])
    const [isFilterApplied, updateIsFilterApplied] = useState(false)
    const {countries} = useCountryState()

    useEffect(() => {
        const fetchData = async (): Promise<any> => {
            const response = await fetch(`https://randomuser.me/api/?results=30&page=${page}&nat=${countries.join(',')}`)
            const data = await response.json()
            if (data.results.length === 0)
                updateHasMoreUsers(false)
            else
                updateUserList(userList.concat(data.results))
        }
        fetchData()
    }, [page]);

    const onSubmit = (userSearch: UserSearchInterface) => {
        const result = filterUserByFirstNameAndLastName(userSearch.name, userList)
        updateIsFilterApplied(true)
        updateFilteredUserList(result)
    }

    const headers = [
        {
            name: "thumbnail",
            label: "Profile Image",
            customComponent: (value: string)=> (<img className='thumbnail' src={value}/>)
        },
        {
            name: "firstName",
            label: "First Name"
        },
        {
            name: "lastName",
            label: "Last Name"
        },
        {
            name: "email",
            label: "Email"
        },
        {
            name: "gender",
            label: "Gender"
        },
        {
            name: "nat",
            label: "Nationality"
        },
        {
            name: "country",
            label: "Location"
        }
    ]
    const usersToShow = isFilterApplied ? filteredUserList : userList

    const renderTable = () => (
        <Table header={headers} data={usersToShow.map((user) => {
            return {
                thumbnail: user.picture.thumbnail,
                email: user.email,
                gender: user.gender,
                nat: user.nat,
                firstName: user.name.first,
                lastName: user.name.last,
                country: user.location.country,
            }
        })}/>
    )

    const renderInfiniteScrollTable = () => (
        <InfiniteScroll
            dataLength={userList.length} //This is important field to render the next data
            next={() => {
                updatePage(page + 1)
            }}
            height={'60vh'}
            hasMore={hasMoreUsers}
            loader={<h4 className='text-align-center'>Loading...</h4>}
            endMessage={
                <p className='text-align-center'>
                    <b>Yay! You have seen it all</b>
                </p>
            }
        >
            {renderTable()}
        </InfiniteScroll>
    )

    return (
        <main className='main-page'>
            <section className='main-page__search'>
                <h3>Search Users</h3>
                <UserSearch onSubmit={onSubmit} onClearFilter={() => updateIsFilterApplied(false)}/>
            </section>
            <section className='main-page__users-list'>
                <h3>Users {isFilterApplied && <i> ( pagination doesn't work with search filter )</i>}</h3>
                {isFilterApplied ? renderTable() : renderInfiniteScrollTable()}
            </section>
        </main>
    );
}

export default HomePage;

