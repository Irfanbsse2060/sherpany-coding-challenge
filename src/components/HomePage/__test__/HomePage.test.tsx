import React from "react";
import {fireEvent, render, wait} from '@testing-library/react'
import HomePage from '../index'
import {CountryProvider} from "../../Context/CountryContext";

import users from "../../../utils/fixtures/userfixture";


describe('Homepage', () => {
    beforeEach(() => {
        fetch.resetMocks();
        fetch.mockResponseOnce(JSON.stringify({results: users}));
    });

    const renderComponent = () => {
        return (
            <CountryProvider>
                <HomePage/>
            </CountryProvider>
        )
    }

    test('should be able to render', async () => {
        const {container, getAllByTestId} = render(renderComponent())
        await wait(() => expect(getAllByTestId('custom-table-data-rows')).toHaveLength(users.length))
        expect(container).toMatchSnapshot()
    });

    test('should be able to search the users by name', async () => {
        const searchName = users[0].name.first
        const {getAllByTestId, getByTestId, getByText} = render(renderComponent())
        await wait(() => expect(getAllByTestId('custom-table-data-rows')).toHaveLength(users.length))

        fireEvent.change(getByTestId('text_field_name'), {target: {value: searchName}})
        fireEvent.click(getByTestId('submit-bt'))

        expect(getAllByTestId('custom-table-data-rows')).toHaveLength(1)
        getByText(searchName)
        getByText(users[0].email)
    });

})
