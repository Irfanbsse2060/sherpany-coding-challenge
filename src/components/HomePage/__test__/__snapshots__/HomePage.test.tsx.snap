// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Homepage should be able to render 1`] = `
<div>
  <main
    class="main-page"
  >
    <section
      class="main-page__search"
    >
      <h3>
        Search Users
      </h3>
      <form
        class="user-search"
      >
        <div
          class="text-field"
        >
          <label
            class="text-field__label"
          >
            Name
          </label>
          <input
            class="text-field__input"
            data-testid="text_field_name"
            name="name"
            required=""
            type="text"
            value=""
          />
        </div>
        <div
          class="user-search__action"
        >
          <button
            class="custom-button custom-button--primary"
            data-testid="submit-bt"
          >
            Search
          </button>
          <button
            class="custom-button custom-button--secondary"
            data-testid="clear-bt"
          >
            Clear search
          </button>
        </div>
      </form>
    </section>
    <section
      class="main-page__users-list"
    >
      <h3>
        Users
      </h3>
      <div
        class="infinite-scroll-component__outerdiv"
      >
        <div
          class="infinite-scroll-component "
          style="height: 60vh; overflow: auto;"
        >
          <table
            class="custom-table"
          >
            <thead>
              <tr
                class="custom-table__row"
              >
                <th
                  class="custom-table__row--header-column"
                >
                  Profile Image
                </th>
                <th
                  class="custom-table__row--header-column"
                >
                  First Name
                </th>
                <th
                  class="custom-table__row--header-column"
                >
                  Last Name
                </th>
                <th
                  class="custom-table__row--header-column"
                >
                  Email
                </th>
                <th
                  class="custom-table__row--header-column"
                >
                  Gender
                </th>
                <th
                  class="custom-table__row--header-column"
                >
                  Nationality
                </th>
                <th
                  class="custom-table__row--header-column"
                >
                  Location
                </th>
              </tr>
            </thead>
            <tbody>
              <tr
                class="custom-table__row"
                data-testid="custom-table-data-rows"
              >
                <td
                  class="custom-table__row--body-column"
                >
                  <img
                    class="thumbnail"
                    src="https://randomuser.me/api/portraits/thumb/men/98.jpg"
                  />
                </td>
                <td
                  class="custom-table__row--body-column"
                >
                  Lachlan
                </td>
                <td
                  class="custom-table__row--body-column"
                >
                  Roberts
                </td>
                <td
                  class="custom-table__row--body-column"
                >
                  lachlan.roberts@example.com
                </td>
                <td
                  class="custom-table__row--body-column"
                >
                  male
                </td>
                <td
                  class="custom-table__row--body-column"
                >
                  NZ
                </td>
                <td
                  class="custom-table__row--body-column"
                >
                  New Zealand
                </td>
              </tr>
              <tr
                class="custom-table__row"
                data-testid="custom-table-data-rows"
              >
                <td
                  class="custom-table__row--body-column"
                >
                  <img
                    class="thumbnail"
                    src="https://randomuser.me/api/portraits/thumb/women/4.jpg"
                  />
                </td>
                <td
                  class="custom-table__row--body-column"
                >
                  Leona
                </td>
                <td
                  class="custom-table__row--body-column"
                >
                  Campbell
                </td>
                <td
                  class="custom-table__row--body-column"
                >
                  leona.campbell@example.com
                </td>
                <td
                  class="custom-table__row--body-column"
                >
                  female
                </td>
                <td
                  class="custom-table__row--body-column"
                >
                  AU
                </td>
                <td
                  class="custom-table__row--body-column"
                >
                  Australia
                </td>
              </tr>
              <tr
                class="custom-table__row"
                data-testid="custom-table-data-rows"
              >
                <td
                  class="custom-table__row--body-column"
                >
                  <img
                    class="thumbnail"
                    src="https://randomuser.me/api/portraits/thumb/women/72.jpg"
                  />
                </td>
                <td
                  class="custom-table__row--body-column"
                >
                  Emilie
                </td>
                <td
                  class="custom-table__row--body-column"
                >
                  Nielsen
                </td>
                <td
                  class="custom-table__row--body-column"
                >
                  emilie.nielsen@example.com
                </td>
                <td
                  class="custom-table__row--body-column"
                >
                  female
                </td>
                <td
                  class="custom-table__row--body-column"
                >
                  DK
                </td>
                <td
                  class="custom-table__row--body-column"
                >
                  Denmark
                </td>
              </tr>
              <tr
                class="custom-table__row"
                data-testid="custom-table-data-rows"
              >
                <td
                  class="custom-table__row--body-column"
                >
                  <img
                    class="thumbnail"
                    src="https://randomuser.me/api/portraits/thumb/women/49.jpg"
                  />
                </td>
                <td
                  class="custom-table__row--body-column"
                >
                  Ester
                </td>
                <td
                  class="custom-table__row--body-column"
                >
                  Woll
                </td>
                <td
                  class="custom-table__row--body-column"
                >
                  ester.woll@example.com
                </td>
                <td
                  class="custom-table__row--body-column"
                >
                  female
                </td>
                <td
                  class="custom-table__row--body-column"
                >
                  DE
                </td>
                <td
                  class="custom-table__row--body-column"
                >
                  Germany
                </td>
              </tr>
            </tbody>
          </table>
          <h4
            class="text-align-center"
          >
            Loading...
          </h4>
        </div>
      </div>
    </section>
  </main>
</div>
`;
